libcontainer/configs/validate/validator.go

```diff
@@ -108,6 +110,27 @@ func (v *ConfigValidator) usernamespace(config *configs.Config) error {
 		if _, err := os.Stat("/proc/self/ns/user"); os.IsNotExist(err) {
 			return fmt.Errorf("USER namespaces aren't enabled in the kernel")
 		}
+
+		root, err := os.Stat("/")
+		if err != nil {
+			return fmt.Errorf("Cannot stat root to check if the process is inside a chroot")
+		}
+
+		// Verify that the current process is not inside a chroot,
+		// the kernel does not allow creating a new user namespace in a chroot.
+		// for information, see: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/kernel/user_namespace.c
+		if root.Sys().(*syscall.Stat_t).Ino != 2 {
```

Comment:
From memory, this check will break if you're trying to nest containers or have any other form of weird rootfs mount structure set up.

Also, I just checked and on my system and `/` appears to have an inode number of `256` with `btrfs`:

```
% stat /
  File: /
  Size: 218             Blocks: 0          IO Block: 4096   directory
Device: 2dh/45d Inode: 256         Links: 1
Access: (0755/drwxr-xr-x)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2017-11-12 07:05:28.393935000 +1100
Modify: 2017-04-27 22:57:13.841679200 +1000
Change: 2017-04-27 22:57:13.841679200 +1000
 Birth: -
```

Reply:
Oh, also this trick won't be able to detect if you have mounted a filesystem onto `/jail` and then done `chroot /jail` -- because then the inode number will still be the same.

There are some ways to detect `chroot` "sort of", using a combination of `/proc/$pid/root` and `/proc/1/root`, or `/proc/1/mountinfo`. But unfortunately they're all very iffy and don't handle all of the corner cases. It would be ideal if the kernel exposed [`current_chrooted`](http://elixir.free-electrons.com/linux/v4.14/source/fs/namespace.c#L3319) to userspace but I imagine that's not likely to happen. Another possibility is to propose that we should be able to introspect the root of `/proc/$pid/ns/mnt` -- something that Eric is more likely to consider (but passing paths around is generally not a good idea especially since the kernel likes to changes paths to `/` if it feels that something fishy is going on).

Reply:
@cyphar Tried the pivot_root approach however doesn't seem to be very reliable. Thanks for taking the time to look at this